internal/ethereum/confirmation_reconciler.go

```diff
@@ -0,0 +1,259 @@
+	txBlockNumber := txBlockInfo.BlockNumber.Uint64()
+	targetBlockNumber := txBlockInfo.BlockNumber.Uint64() + targetConfirmationCount
+
+	// Check if the canonical chain has caught up to the transaction block
+	chainTail := bl.canonicalChain.Back().Value.(*ffcapi.MinimalBlockInfo)
+	if chainTail == nil || chainTail.BlockNumber.Uint64() < txBlockNumber {
+		log.L(ctx).Debugf("Canonical chain is waiting for the transaction block %d to be indexed", txBlockNumber)
+		return nil, nil, true
+	}
+
+	// Build new confirmations from blocks after the transaction block
+
+	newConfirmationsWithoutTxBlock = []*ffcapi.MinimalBlockInfo{}
+	currentBlock := bl.canonicalChain.Front()
+	for currentBlock != nil {
+		currentBlockInfo := currentBlock.Value.(*ffcapi.MinimalBlockInfo)
+
+		// If we've reached the target confirmation count, mark as confirmed
+		if currentBlockInfo.BlockNumber.Uint64() > targetBlockNumber {
+			reconcileResult.Confirmed = true
+			// if the canonical chain contains the next block after the target block number,
+			// and the new confirmations queue is empty,
+			// we set the last validated block to the next block, so the downstream function can use it validate blocks before it
+			if len(newConfirmationsWithoutTxBlock) == 0 && currentBlockInfo.BlockNumber.Uint64() == targetBlockNumber+1 {
+				lastValidatedBlock = currentBlockInfo
+			}
+			break
+		}
+
+		// Skip blocks at or before the transaction block
+		if currentBlockInfo.BlockNumber.Uint64() <= txBlockNumber {
+			currentBlock = currentBlock.Next()
+			continue
+		}
+
+		// Add blocks after the transaction block to confirmations
+		newConfirmationsWithoutTxBlock = append(newConfirmationsWithoutTxBlock, &ffcapi.MinimalBlockInfo{
+			BlockHash:   currentBlockInfo.BlockHash,
+			BlockNumber: fftypes.FFuint64(currentBlockInfo.BlockNumber.Uint64()),
+			ParentHash:  currentBlockInfo.ParentHash,
+		})
+		currentBlock = currentBlock.Next()
+	}
+	return newConfirmationsWithoutTxBlock, lastValidatedBlock, false
+}
+
+// initializeConfirmationMap initializes the confirmation map with the transaction block
+// and validates existing confirmations against the current transaction block.
+// Returns existing confirmations if valid, or nil if a fork is detected.
+func (bl *blockListener) initializeConfirmationMap(reconcileResult *ffcapi.ConfirmationMapUpdateResult, txBlockInfo *ffcapi.MinimalBlockInfo) []*ffcapi.MinimalBlockInfo {
+	// If no existing confirmations, initialize with the transaction block
+	if len(reconcileResult.Confirmations) == 0 {
+		reconcileResult.Confirmations = []*ffcapi.MinimalBlockInfo{txBlockInfo}
+		return nil
+	}
+
+	// Validate existing confirmations against the current transaction block
+	existingQueue := reconcileResult.Confirmations
+	if len(existingQueue) > 0 {
+		existingTxBlock := existingQueue[0]
+		if !existingTxBlock.Equal(txBlockInfo) {
+			// Transaction block mismatch indicates a fork - rebuild confirmation queue
+			reconcileResult.NewFork = true
```

Comment:
Just to ensure we're not considering this the only thing that is a fork.

Forks when they happen, aren't assured to affect the first block in the chain we care about - they might only affect the chain after that (from index `1` or index `5`). In fact if you think about a list of confirmations like 50 - it's way more likely for forks to affect the later parts, than the start.

If we mean `DifferentTransactionBlock` we should use that terminology or similar.

If we mean `Fork` - then we should treat it like the description I put above.